packages/qwik/src/core/util/string.ts

```diff
@@ -0,0 +1,128 @@
+// pack bytes into valid UTF-16 string
+//
+// strategy:
+//
+// * using esc as the escape character
+//  * if there is esc in the bytes, double it
+// * if there is unmatched surrogate pair, mark it by the escape character
+//
+// 0x007f: escape, because it's rare but still only one utf-8 byte.
+//   To escape itself, use 0x007f 0x08ff (two bytes utf-8)
+// 0x0000->0x001f: converted to esc + 0x0020->0x003f (two bytes utf-8)
+// unmatched pairs: converted to esc + (code-0xd800+0x0040) 0x0040->0x083f
+//   (two-four bytes utf-8)
+// BOM: esc + 0x08fe (four bytes utf-8)
+//
+// If the length of the bytes is odd, the last byte XX is put after the escape
+// character as 0xFFXX.
+//
+const ESC = 0x007f;
+const SURROGATE_OFFSET = 0xd800 - 0x0040;
+export const packUint8Array = (bytes: Uint8Array) => {
+  const odd = bytes.length % 2 === 1;
+  const dbytes = new Uint16Array(bytes.buffer, 0, bytes.length >> 1);
+  let code = '';
+  let surrogate = false;
+  for (let i = 0; i < dbytes.length; i++) {
+    const c = dbytes[i];
+    // test high surrogate
+    if (c >= 0xd800 && c <= 0xdbff) {
+      if (surrogate) {
+        // unmatched high surrogate
+        // omit the MSB to make it a normal codepoint
+        const prev = dbytes[i - 1];
+        code += String.fromCharCode(ESC, prev - SURROGATE_OFFSET);
+      }
+      surrogate = true;
+      continue;
+    }
+    // test low surrogate
+    if (c >= 0xdc00 && c <= 0xdfff) {
+      if (surrogate) {
+        // valid surrogate pair
+        code += String.fromCharCode(dbytes[i - 1], c);
+        surrogate = false;
+        continue;
+      }
+      // unmatched low surrogate
+      // omit the MSB to make it a normal codepoint
+      code += String.fromCharCode(ESC, c - SURROGATE_OFFSET);
+      continue;
+    }
+    if (surrogate) {
+      // no low surrogate after high surrogate
+      // omit the MSB to make it a normal codepoint
+      const prev = dbytes[i - 1];
+      code += String.fromCharCode(ESC, prev - SURROGATE_OFFSET);
+      surrogate = false; // reset surrogate
+    }
+    // escape the BOM
+    if (c === 0xfeff) {
+      // BOM
+      code += String.fromCharCode(ESC, 0x08fe);
+      continue;
+    }
+    // extra comaction againt the strinfigy of the control characters
+    if (c <= 0x001f) {
+      code += String.fromCharCode(ESC, c + 0x0020);
+      continue;
+    }
+    // double the escape character
+    if (c === ESC) {
+      code += String.fromCharCode(0x08ff);
+    }
+    // normal codepoint
+    code += String.fromCharCode(c);
+  }
+  if (surrogate) {
+    // ended with unmatched high surrogate
+    // subtract the offset to make it a normal codepoint
+    const c = dbytes[dbytes.length - 1];
+    code += String.fromCharCode(ESC, c - SURROGATE_OFFSET);
+  }
+  if (odd) {
+    // put the last byte
+    code += String.fromCharCode(ESC, bytes[bytes.length - 1] + 0x0900);
```

Comment:
Since this can happen often, perhaps we should just start the output with O or E? Only one byte.

Reply:
The UTF-16 string, JavaScript using internally, must be even length in bytes, so if the Uint8Array is odd length, always one extra byte padding is needed.
Current approach is add ESC(1 byte) + 1 character for odd ones and none for even ones.
If we put the O or E, it costs `O` (1 byte) for odd ones and `E` + `padding` (2 bytes) for even ones.
So I think they are mostly same efficiency.

Reply:
In addition, I think the length of the buffers tend to be even byte length.

Reply:
Yes but it turns a single byte into 4 bytes...

Reply:
Hmm you're right, most likely they're even.

Reply:
So the algorithm would be, go through the uint8 two bytes at a time, and if odd encode the last byte as 0x0000+byte and add esc.

On decode just check if the final byte is esc, then the last charcode does not add the high byte.

Reply:
0x0000+byte means 0x00XX where XX is the byte?
If so, that isn't the swap the position of the ESC and the byte, of current algorithm?

Reply:
Changed the implementation to use `0x00XX ESC` for the last byte.
It was needed to take a special treatment for `0x007f 0x007f` coming at the end to avoid the collision.

Reply:
Found an issue. I am fixing....